cdap-proto/src/main/java/io/cdap/cdap/proto/TriggeringInfo.java

```diff
@@ -0,0 +1,159 @@
+ */
+public abstract class TriggeringInfo implements Trigger {
+  private final Type type;
+  private final ScheduleId scheduleId;
+  /**
+   * Runtime Arguments propagated from upstream program or properties of time schedule
+   */
+  @Nullable
+  private final Map<String, String> runtimeArguments;
+
+  protected TriggeringInfo(Type type, ScheduleId scheduleId, @Nullable Map<String, String> runtimeArguments) {
+    this.type = type;
+    this.scheduleId = scheduleId;
+    this.runtimeArguments = runtimeArguments;
+  }
+
+  @Override
+  public Type getType() {
+    return type;
+  }
+
+  public ScheduleId getScheduleId() {
+    return scheduleId;
+  }
+
+  @Nullable
+  public Map<String, String> getRuntimeArguments() {
+    return runtimeArguments;
+  }
+
+  public static class TimeTriggeringInfo extends TriggeringInfo {
+    private final String cronExpression;
+
+    public TimeTriggeringInfo(ScheduleId scheduleId, Map<String, String> runtimeArguments, String cronExpression) {
+      super(Type.TIME, scheduleId, runtimeArguments);
+      this.cronExpression = cronExpression;
+    }
+
+    public String getCronExpression() {
+      return cronExpression;
+    }
+  }
+
+  public static class ProgramStatusTriggeringInfo extends TriggeringInfo {
+    private final ProgramRunId programRunId;
+
+    public ProgramStatusTriggeringInfo(ScheduleId scheduleId, Map<String, String> runtimeArguments,
+                                          ProgramRunId programRunId) {
+      super(Type.PROGRAM_STATUS, scheduleId, runtimeArguments);
+      this.programRunId = programRunId;
+    }
+
+    public ProgramRunId getProgramRunId() {
+      return programRunId;
+    }
+  }
+
+  public static class PartitionTriggeringInfo extends TriggeringInfo {
+    private final String datasetNamespace;
+    private final String datasetName;
+    private final int expectedNumPartitions;
+    private final int actualNumPartitions;
+
+
+    public PartitionTriggeringInfo(ScheduleId scheduleId, @Nullable Map<String, String> runtimeArguments,
+                                      String datasetName, String datasetNamespace, int expectedNumPartitions,
+                                      int actualNumPartitions) {
+      super(Type.PARTITION, scheduleId, runtimeArguments);
+      this.datasetName = datasetName;
+      this.datasetNamespace = datasetNamespace;
+      this.expectedNumPartitions = expectedNumPartitions;
+      this.actualNumPartitions = actualNumPartitions;
+    }
+
+    public String getDatasetNamespace() {
+      return datasetNamespace;
+    }
+
+    public String getDatasetName() {
+      return datasetName;
+    }
+
+    public int getExpectedNumPartitions() {
+      return expectedNumPartitions;
+    }
+
+    public int getActualNumPartitions() {
+      return actualNumPartitions;
+    }
+  }
+
+  public abstract static class AbstractCompositeTriggeringInfo<T extends TriggeringInfo> extends TriggeringInfo {
+    private final List<T> triggeringInfos;
```

Comment:
nit: Does this have to be generic? The current subclasses are using TriggeringInfo, so why not use `List<TriggeringInfo>` ? Is there a restriction on the types in composite triggers?